server/interceptor/interceptor.go

```diff
@@ -413,6 +413,28 @@ type logLeavesResponse interface {
 	GetQueuedLeaves() []*trillian.QueuedLogLeaf
 }
 
+// SlowRPC returns an interceptor which will call f if an RPC takes longer than deadline to complete.
+func SlowRPC(deadline time.Duration, f func()) grpc.UnaryServerInterceptor {
+	s := slowRPC{
+		f:        f,
+		deadline: deadline,
+	}
+	return s.UnaryServerInterceptor
+}
+
+type slowRPC struct {
+	f        func()
+	deadline time.Duration
+}
+
+// UnaryServerInterceptor executes the SlowRPC logic for unary RPCs.
+func (s *slowRPC) UnaryServerInterceptor(ctx context.Context, req interface{}, _ *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	t := time.AfterFunc(s.deadline, s.f)
+	resp, err := handler(ctx, req)
+	t.Stop()
```

Comment:
The documentation for AfterFunc:

"AfterFunc waits for the duration to elapse and then calls f in its own goroutine. It returns a Timer that can be used to cancel the call using its Stop method."

makes it sound like there's no guarantee here that s.f will run to completion.  As it's run in its own goroutine it looks like if s.f takes longer to run than handler, it will only run until handler has completed, and will then be stopped by the t.Stop.

Maybe this is fine? I wanted to check that this is WAI, as these functions are part of the package API, so could be imported by anyone and used in a way you may not have thought of.  If it is the case that that is what you intend, it might be worth noting in the comments for either UnaryServerInterceptor or SlowRPC or both - particularly as the current function comment for SlowRPC implies that f will be called after the deadline, but does not mention anywhere that it may not run to completion.

Reply:
I think this works:

"For a timer created with AfterFunc(d, f), if t.Stop returns false, then the timer has already expired and the function f has been started in its own goroutine; Stop does not wait for f to complete before returning. "